Improve fake tensor leakage detection in export by not relying on gc too much

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #163516

Previously we relied on gc to get the snapshot of fake tensors before and after export to get list of fake tensors that are created during export. This caused some flakiness in our test suite (https://github.com/pytorch/pytorch/issues/162232). it seems super hard to make gc deterministic, so we just instrument fake tensor creation which seems lot better. In addition, it is also quite faster than previous approach becuase we are no longer manually triggering garbage collector. 


cc @ezyang @EikanWang @jgong5 @wenzhe-nrv

Differential Revision: [D82966648](https://our.internmc.facebook.com/intern/diff/D82966648)

tugsbayasgalan committed:
Improve fake tensor leakage detection in export by not relying on gc too much

[ghstack-poisoned]

ezyang reviewed (approved):
sure why not

ezyang commented:
cc @eellison for fake tensor constructor modifications

eellison reviewed (commented):
why do we track constructors specifically ? We already track every fake tensor in the fake tensor converter.

tugsbayasgalan committed:
Update on "Improve fake tensor leakage detection in export by not relying on gc too much"


Previously we relied on gc to get the snapshot of fake tensors before and after export to get list of fake tensors that are created during export. This caused some flakiness in our test suite (https://github.com/pytorch/pytorch/issues/162232). it seems super hard to make gc deterministic, so we just instrument fake tensor creation which seems lot better. In addition, it is also quite faster than previous approach becuase we are no longer manually triggering garbage collector. 


cc ezyang EikanWang jgong5 wenzhe-nrv

Differential Revision: [D82966648](https://our.internmc.facebook.com/intern/diff/D82966648)

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge -f "Landed internally"
